### DISCUSSION TITLE
Typing for subforms/complex fields

### DISCUSSION THREAD

**Discussion by abelgibayev** [react-hook-form/react-hook-form#6525] [Q&A]
**Is your feature request related to a problem? Please describe.**
I'm migrating an application from formik to react-hook-form and have a problem with migrating complex fields. The idea of them is reusing group of fields in different forms. E.g. Sport and Player entities have field 'translations' that is working with the same subform. This becomes hard if I want to use this subform in different entities, once it may be just `'translations'` in other cases `'player.translations'`

**Describe the solution you'd like**
I'm using `useController` hook to work with this field and would like to define types of this field inside subform, like
`const controller = useController<Translations[]>({ name })` like it's in formik.

**Describe alternatives you've considered**
I tried to accept generic type `FormValues extends {translations: Translation[]}`, but the default value doesn't accept nor `[]`, nor `[] as Translation[]`, even if the name prop of the hook is clearly set as `'translations'`

**Additional context**
![image](https://user-images.githubusercontent.com/22354231/132511397-4d8d0575-cbbe-469e-84f6-69a30766650c.png)

I tried to find answers in documentations, but typescript is used only in `useForm` hook in all examples.

**Comment by bluebill1049** [comment]
can you provide a codesandbox for the above usage?

**Comment by abelgibayev** [comment]
[sandbox](https://codesandbox.io/s/hungry-kapitsa-u1mfh)

**Comment by bluebill1049** [comment]
can you provide some details for your codesandbox above and remove all those extra codes if you can? thanks

**Reply by imballinst** [reply]
@bluebill1049 apologies for joining in the discussion—this is also something that I am interested in. Say, I have a subform with only `firstName` and `lastName`. Then, I want to use both fields in a new form, say, `middleName`. Would it be possible that there can be a "generic" component that can render `firstName` and `lastName`?

Example sandbox: https://codesandbox.io/s/admiring-brook-b1qyq?file=/src/App.tsx. What I want to do is that, is there a way so that `IsolateReRenderEntireForm` can fit a `control` for `Subform` and `EntireForm`?

EDIT: I just did some more searches and found out https://github.com/react-hook-form/react-hook-form/discussions/4807#discussioncomment-614502. I guess it won't be possible to do it, then.
EDIT2: probably using `extends Subform` works (a bit hacky), then in the component I asserted its type: https://codesandbox.io/s/cranky-bell-ugmxv?file=/src/App.tsx. This ensures that the `IsolateReRenderEntireForm` is still properly checked when we write the JSX and inside the component, we can still get the actual fields from `useWatch`.

**Reply by bluebill1049** [reply]
is this what you after? https://github.com/react-hook-form/react-hook-form/pull/6562

**Reply by imballinst** [reply]
Ooo! That's a very interesting PR, yes, I think that's what I am looking for. Thanks for referencing it! It raises a question, though, what's the difference between:

```ts
useController({ control, name: 'test' }).field.value
```

and

```ts
useWatch({ control, name: 'test' })
```

Assuming `test` is a field for string input, when do we use `useController` and when do we use `useWatch`? Is it always `useWatch` when it comes to _only_ getting the value?
